File: Algorithm/Code_Materials/c_or_cpp/charpter9_binary/level2/missing_number.cpp
#include <stdio.h>

 
int missingNumber(int* a, int length) {
  int left = 0;
  int right = length - 1;
  while (left <= right) {
    int mid = (left + right) / 2;
    if (a[mid] == mid) {
    		left = mid + 1;
    } else {
   		 right = mid - 1;
    }
  }
  return left;
}

int main()
{
    int array[] = {0,1, 2, 3, 5,6,7};
    int index = missingNumber(array, 7);
    if (index == -1)
    {
        printf("Target not found in the array.\n");
    }
    else
    {
        printf("Target found at index %d.\n", index);
    }
    return 0;
}

File: Algorithm/Leetcode/Code_Materials/Java/src/main/java/com/yugutou/charpter9_bitree_bisearch/level2/MissingNumber.java
package com.yugutou.charpter9_bitree_bisearch.level2;

/**
 * 剑指 Offer 53 - II. 0～n-1中缺失的数字
 */
public class MissingNumber {
    public static void main(String[] args) {
        int[] nums = {0, 1, 2, 3, 4, 5, 6, 7, 9};
        System.out.println(solve(nums));
    }

    public static int solve(int[] a) {
        // write code here
        int left = 0;
        int right = a.length - 1;
        while (left < right) {
            int mid = (left + right) / 2;
            if (a[mid] == mid) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}
